datasize_macro/src/impl_data_sizes.rs

```diff
@@ -0,0 +1,175 @@
+					}
+				}
+				Fields::Unnamed(f) => {
+					// Set a name for all fields of an unnamed variant
+					let names: Vec<Ident> = f
+						.unnamed
+						.iter()
+						.enumerate()
+						.map(|(i, _)| format_ident!("f{}", i))
+						.collect();
+
+					quote! {
+						Self::#ident (#(#names),*) => { usize::default() #( + #names.data_size())*},
+					}
+				}
+				// If the variant is a unit variant, then the size is 0
+				Fields::Unit => quote! {
+					Self::#ident => 0,
+				},
+			}
+		})
+		.fold(quote!(), |t, b| quote! (#t #b));
+
+	quote! {
+		match &self {
+			#branches
+		}
+	}
+}
+
+fn impl_datasize_struct(data_struct: &DataStruct) -> TokenStream2 {
+	match &data_struct.fields {
+		Fields::Named(f) => {
+			// Get a list of all named fields of a named variant
+			let names: Vec<Ident> = f.named.iter().filter_map(|f| f.ident.to_owned()).collect();
+
+			quote! {
+				usize::default() #(+ self.#names.data_size())*
+			}
+		}
+		Fields::Unnamed(f) => {
+			// Set a name for all fields of an unnamed variant
+			let names: Vec<Index> = f
+				.unnamed
+				.iter()
+				.enumerate()
+				.map(|(i, _)| Index::from(i))
+				.collect();
+
+			quote! {
+				usize::default() #(+ self.#names.data_size())*
+			}
+		}
+		// If the variant is a unit variant, then the size is 0
+		Fields::Unit => quote!(usize::default()),
+	}
+}
+
+fn impl_static_datasize_enum(data_enum: &DataEnum) -> TokenStream2 {
+	data_enum
+		.variants
+		.iter()
+		.map(|v| {
+			// Retrieve types for all fields
+			let types: Vec<Type> = v.fields.iter().map(|f| f.ty.to_owned()).collect();
+			quote! {
+				usize::default() #( + #types::static_data_size())*
+			}
+		})
+		// Use the maximum size among all variants
+		.fold(
+			quote!(usize::default()),
+			|t, b| quote!(std::cmp::max(#t, #b)),
+		)
+}
+
+fn impl_static_datasize_struct(data_struct: &DataStruct) -> TokenStream2 {
+	// Retrieve types of all fields
+	let types: Vec<TokenStream2> = data_struct
+		.fields
+		.iter()
+		.map(|f| f.ty.to_owned())
+		.map(replace_type_syntax)
+		.collect();
+
+	// We call `static_data_size()` on each of the names
+	quote! ( usize::default() #(+ <#types>::static_data_size())*)
+}
+
+/// This replaces all types to a syntax on which we can call functions
+fn replace_type_syntax(t: Type) -> TokenStream2 {
+	match t {
+		Type::Tuple(t) => {
+			let types: Vec<TokenStream2> = t
+				.elems
+				.iter()
+				.map(|t| replace_type_syntax(t.to_owned()))
+				.collect();
+			quote!((#(#types),*))
+		}
+		Type::Ptr(p) => {
+			let ty = replace_type_syntax(*p.elem);
+			quote!(#ty)
+		}
+		Type::Array(a) => {
+			let ty = replace_type_syntax(*a.elem);
+			quote!([#ty])
+		}
+		Type::Slice(s) => {
+			let ty = replace_type_syntax(*s.elem);
+			quote!([#ty])
+		}
+		Type::Reference(r) => {
+			let ty = replace_type_syntax(*r.elem);
+			quote!(#ty)
+		}
+		Type::Infer(i) => i.to_token_stream(),
+		Type::Path(p) => p.to_token_stream(),
+		_ => panic!("This type is not supported yet"),
+	}
+}
```

Comment:
Well i found out it was possible [here](https://github.com/XdotRS/cornflakes/pull/1/commits/a4b028c0b6594b18dfb4d70cc066c556a4ea840f) and i removed what I have done, but kept that for calls with references.
It just call on the type inside the reference (or raw pointer).
I think it could be resolve by implementing `StaticDataSize` for references too

Reply:
Yes, the data size traits should be implemented for references, mutable references, boxes probably, as just being the same as the dereferenced thing... I wonder if it's possible to implement for any generic type `R` that can be dereferenced to `T` where `T` is the data size trait implementing type, not sure (as in, so it can be implemented for any kind of smart pointer, not that it matters because nothing other than shared references, mutable references, and boxes is likely to be used, but it would be nice to be generic if that is actually possible).

Reply:
I pretty sure you can call `data_size()` on any referenced type implementing `DataSize` at runtime, even if it is not explicitly implemented for the reference.
The only thing needed is implementing `StaticDataSize` for all references, which I have just done